### ISSUE TITLE
Automatic deriving for ToJson instance

### ISSUE THREAD

**Issue by sanketr** [chadaustin/buffer-builder#11]
Is it trivial to adapt `deriveJson` from `Aeson` to automatically derive `ToJson` instances? I looked into using buffer builder for a project but got stopped in tracks because I use automatic derivation everywhere (instead of hand-coding it tediously). An example derivation is like this through `Data.Aeson.TH`:
`deriveJSON defaultOptions{fieldLabelModifier = Prelude.drop 5,constructorTagModifier = Prelude.map toLower,omitNothingFields = True} ''ARecordType`

It seems that in order to use `encodeJson`, `ToJson` instance must be provided which must be hand-coded for now. If it can somehow be derived from `Data.Aeson.ToJSON`, then it will make it very useful! 

Unless I missed something about how to adapt an existing Aeson `ToJSON` to `ToJson`?

**Comment by chadaustin** [comment]
This is a nice feature request and would be logical to add to the library.

For now, though, you could get some of the speedup without changing all of your code by trying the [buffer-builder-aeson](https://hackage.haskell.org/package/buffer-builder-aeson) package.  Have you tried that?

**Comment by sanketr** [comment]
Yeah, I looked into buffer-builder-aeson before but wasn't sure what to make of it performance-wise. So, from performance perspective, would an encoding call like below be performance-wise almost as good as hand-writing the instance from scratch?
`encodeJson . toJSON`

 For example, instead of 400% speed-up with hand-written `toJson` instance, if we have like 390% speedup through buffer-builder-aeson, that is very good in my books.

If that is the case, then I guess this request is redundant and unnecessary. We can use all of Aeson machinery and just swap in buffer-builder using the call above, instead of `encode` to get most of the speed-up.

**Comment by sanketr** [comment]
Hmm, I just ran a benchmark using your benchmark script and a 29KB test.json. I modified it to add `encodeJson . toJSON` as `bufferbuilder2`. There is speed-up compared to `Aeson` but not orders of magnitudes like `bufferbuilder`. Output below:

```
benchmarking render/bufferbuilder
time                 140.8 μs   (140.5 μs .. 141.2 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 141.0 μs   (140.8 μs .. 141.5 μs)
std dev              1.114 μs   (701.8 ns .. 1.802 μs)

benchmarking render/bufferbuilder2
time                 464.9 μs   (463.6 μs .. 466.2 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 465.4 μs   (463.8 μs .. 468.6 μs)
std dev              7.185 μs   (4.075 μs .. 13.03 μs)

benchmarking render/aeson
time                 657.8 μs   (655.3 μs .. 660.4 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 659.1 μs   (657.2 μs .. 660.7 μs)
std dev              5.743 μs   (4.745 μs .. 7.600 μs)
```

**Comment by andyfriesen** [comment]
That makes sense to me.  The main reason buffer-builder is so much faster than Aeson is because it doesn't build any intermediate structure.  It streams your data directly into a bytestring.

I'm totally in favour of adding this feature to the library, but I can't offer any implementation advice because I don't have a lot of experience with GHC generics.

**Comment by chadaustin** [comment]
Thanks for doing the benchmark, @sanketr!

Unfortunately, I'm in the same boat as @andyfriesen.  I've never used GHC generics and while I'd love to learn them I don't have that much spare time right now.

@sanketr Do you want to try taking the Aeson implementation and adapt it to buffer-builder in a PR?

Thanks!

**Comment by sanketr** [comment]
@chadaustin, guess we all are in the same boat then :) I probably will come back to this if Aeson performance becomes a major issue, and take a shot. For the time being, I can just use up more processors because each encoding runs in its own snap websocket thread.

About Generics approach - I have looked at Aeson code before, and if I am not mistaken, it is mostly template haskell for `Data.Aeson.TH` module that we typically use for derivation. Aeson has `Generics` version too but template haskell seems to produce much more performant code (not sure if this is still true), and seems to be used more from what I have seen so far.

**Comment by unhammer** [comment]
Note: It's possible to move to buffer-builder incrementally. We had lots of Aeson instances. On a simple test set, where we write a 20MB json file, we had ~490M memory usage. Simply switching to buffer-builder-aeson without writing any instances brought it down to ~315M. Then I tried just writing instances for the top-level structure, using `instance J.ToJson X where toJson = J.toJson . Aeson.toJSON . oldAesonInstance` almost everywhere except for certain lists I knew to be long. This brought memory usage down to 160M =D
